go/viperutil/config.go

```diff
@@ -0,0 +1,297 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package viperutil
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"vitess.io/vitess/go/viperutil/funcs"
+	"vitess.io/vitess/go/viperutil/internal/log"
+	"vitess.io/vitess/go/viperutil/internal/registry"
+	"vitess.io/vitess/go/viperutil/internal/value"
+)
+
+var (
+	configPaths = Configure(
+		"config.paths",
+		Options[[]string]{
+			GetFunc:  funcs.GetPath,
+			EnvVars:  []string{"VT_CONFIG_PATH"},
+			FlagName: "config-path",
+		},
+	)
+	configType = Configure(
+		"config.type",
+		Options[string]{
+			EnvVars:  []string{"VT_CONFIG_TYPE"},
+			FlagName: "config-type",
+		},
+	)
+	configName = Configure(
+		"config.name",
+		Options[string]{
+			Default:  "vtconfig",
+			EnvVars:  []string{"VT_CONFIG_NAME"},
+			FlagName: "config-name",
+		},
+	)
+	configFile = Configure(
+		"config.file",
+		Options[string]{
+			EnvVars:  []string{"VT_CONFIG_FILE"},
+			FlagName: "config-file",
+		},
+	)
+	configFileNotFoundHandling = Configure(
+		"config.notfound.handling",
+		Options[ConfigFileNotFoundHandling]{
+			Default: WarnOnConfigFileNotFound,
+			GetFunc: getHandlingValue,
+		},
+	)
+	configPersistenceMinInterval = Configure(
+		"config.persistence.min_interval",
+		Options[time.Duration]{
+			Default:  time.Second,
+			EnvVars:  []string{"VT_CONFIG_PERSISTENCE_MIN_INTERVAL"},
+			FlagName: "config-persistence-min-interval",
+		},
+	)
+)
+
+func init() {
+	wd, err := os.Getwd()
+	if err != nil {
+		log.WARN("failed to get working directory (err=%s), not appending to default config-paths", err)
+		return
+	}
+
+	configPaths.(*value.Static[[]string]).DefaultVal = []string{wd}
+	// Need to re-trigger the SetDefault call done during Configure.
+	registry.Static.SetDefault(configPaths.Key(), configPaths.Default())
+}
+
+// RegisterFlags installs the flags that control viper config-loading behavior.
+// It is exported to be called by servenv before parsing flags for all binaries.
+//
+// It cannot be registered here via servenv.OnParse since this causes an import
+// cycle.
+func RegisterFlags(fs *pflag.FlagSet) {
+	fs.StringSlice("config-path", configPaths.Default(), "Paths to search for config files in.")
+	fs.String("config-type", configType.Default(), "Config file type (omit to infer config type from file extension).")
+	fs.String("config-name", configName.Default(), "Name of the config file (without extension) to search for.")
+	fs.String("config-file", configFile.Default(), "Full path of the config file (with extension) to use. If set, --config-path, --config-type, and --config-name are ignored.")
+	fs.Duration("config-persistence-min-interval", configPersistenceMinInterval.Default(), "minimum interval between persisting dynamic config changes back to disk (if no change has occurred, nothing is done).")
```

Comment:
> that setting the in-memory value does not guarantee it persists as this could only be in the process memory, before being written to the FS, and the FS write may not be 

correct

> Do we effectively lose num 2? I guess we could simply document the breaking change in that it's up to the user to reset it back to the default when they're done with their testing (since intermediate temporary test values were likely persisted in the config file). And when choosing to use a config file, you must become aware of this anyway and change your process accordingly. I think that's totally fine and reasonable FWIW.

currently yes. one thing we could do to bring that back is say something like "if `interval` is a negative value then we never persist to disk" but then you run the reverse problem (change `X` in memory, then change `Y` in the file, and when we reload `X` goes back to the original value)

> Precedence also comes into play 

correct, the flag wins if a value is in both spots (including here)

Reply:
also, realizing that

> I would guess that on process normal shutdown all of these pending in-memory changes are flushed? 

we currently don't, but i can do something like:

```go
defer func() {
    select {
        case <-v.setCh:
            persistOnce()
        default:
    }
}()
```

doesn't handle the "killed" case, since `servenv` only provides OnTerm, but also at the end of the day, if the OS decides it's really gonna kill it, and it's not done, well, too bad for us

Reply:
These are all fine. Just considerations that we can internalize and potentially document if needed.